### PR TITLE
Executor

### PR DESCRIPTION
**Pull request by mmamayka**
Adds instruction executor stub

### CODE REVIEW

File: include/besm-666/reg/rf.hpp [hunk 1]
```diff
@@ -6,30 +6,64 @@
...
+    static constexpr const size_t X24 = 24;
+    static constexpr const size_t X25 = 25;
+    static constexpr const size_t X26 = 26;
+    static constexpr const size_t X27 = 27;
+    static constexpr const size_t X28 = 28;
+    static constexpr const size_t X29 = 29;
+    static constexpr const size_t X30 = 30;
+    static constexpr const size_t X31 = 31;
+    static constexpr const size_t PC = 32;
     static constexpr size_t Size = 33;
```
Comment (c71n93) [inline include/besm-666/reg/rf.hpp:50]: why size is not `const`?

File: include/besm-666/exec/executor.hpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    void exec_ADDI(Instruction instr);
+    void exec_SLTI(Instruction instr);
+    void exec_SLTU(Instruction instr);
+    void exec_ANDI(Instruction instr);
+    void exec_ORI(Instruction instr);
+    void exec_XORI(Instruction instr);
+    void exec_LUI(Instruction instr);
+    void exec_AUIPC(Instruction instr);
+
+    reg::GPRF &gprf_;
```
Comment (c71n93) [inline include/besm-666/exec/executor.hpp:25]: Fix refference

File: test/exec/RV64I-executor-tests.cpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    EXPECT_EQ(gprf.read(reg::GPRF::X2), 1);
+}
+TEST_F(RV64IExecutorTest, SLTI_False) {
+    SetupInstrI(InstructionOp::SLTI, reg::GPRF::X2, reg::GPRF::X1, -1);
+    gprf.write(reg::GPRF::X1, 2);
+
+    exec.exec(instr);
+
+    EXPECT_EQ(gprf.read(reg::GPRF::X2), 0);
+}
```
Comment (c71n93) [inline test/exec/RV64I-executor-tests.cpp:65]: Are you planning to add tests to other exec functions

File: src/exec/executor.cpp [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        break;
+    case XORI:
+        exec_XORI(instr);
+        break;
+    case LUI:
+        exec_LUI(instr);
+        break;
+    case AUIPC:
+        exec_AUIPC(instr);
+        break;
```
Comment (c71n93) [inline src/exec/executor.cpp:31]: Add default temporarily

File: test/exec/RV64I-executor-tests.cpp (hunk 3 above)
Comment (mmamayka) [reply test/exec/RV64I-executor-tests.cpp:65]: Of course, they will be added a bit later

File: src/exec/executor.cpp (hunk 4 above)
Comment (mmamayka) [reply src/exec/executor.cpp:31]: Lack of default section makes compiler to show warning with list of unimplemented instructions

File: include/besm-666/reg/rf.hpp (hunk 1 above)
Comment (mmamayka) [reply include/besm-666/reg/rf.hpp:50]: Typo, will be fixed

**Comment by mmamayka** [comment]
Do everybody agree with current executors implementation style?

File: include/besm-666/exec/executor.hpp (hunk 2 above)
Comment (mmamayka) [reply include/besm-666/exec/executor.hpp]: Fixed

File: include/besm-666/reg/rf.hpp (hunk 1 above)
Comment (mmamayka) [reply include/besm-666/reg/rf.hpp]: Fixed

File: src/exec/executor.cpp (hunk 4 above)
Comment (c71n93) [reply src/exec/executor.cpp:42]: > Lack of default section makes compiler to show warning with list of unimplemented instructions

very smart

**Comment by c71n93** [comment]
> Do everybody agree with current executors implementation style?

Yes, style is good. But I checked implementation only of half of instructions.

File: include/besm-666/exec/executor.hpp [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+class Executor {
+public:
+    Executor() = default;
+
+    void exec(Instruction instr);
+
+    inline reg::GPRF const &getState() const { return gprf_; }
+
+private:
+    void exec_ADDI(Instruction instr);
```
Comment (levBagryansky) [inline include/besm-666/exec/executor.hpp:18]: @mmamayka could you plesae mark it const if you removed const from `Finstruction`'s field

File: include/besm-666/instruction.hpp [hunk 6]
```diff
@@ -5,16 +5,16 @@
...
 struct Instruction {
-    const Register rd;
-    const Register rs1;
-    const Register rs2;
-    const Immidiate immidiate;
-    const InstructionOp operation;
+    Register rd;
+    Register rs1;
+    Register rs2;
+    RV64UDWord immidiate;
```
Comment (levBagryansky) [inline include/besm-666/instruction.hpp:16]: @mmamayka use `Immidiate` which you have already reassign

**Review by levBagryansky — CHANGES_REQUESTED** [review]
@mmamayka  Looks good. I'll check tests later

File: include/besm-666/exec/executor.hpp (hunk 5 above)
Comment (mmamayka) [reply include/besm-666/exec/executor.hpp:18]: What to mark?

Comment (levBagryansky) [reply include/besm-666/exec/executor.hpp:18]: @mmamayka I mean `const Instruction instr`. Does `exec_ADDI` change `Executor`?

File: test/exec/RV64I-executor-tests.cpp (hunk 3 above)
Comment (mmamayka) [reply test/exec/RV64I-executor-tests.cpp:512]: Done

File: include/besm-666/exec/executor.hpp (hunk 5 above)
Comment (mmamayka) [reply include/besm-666/exec/executor.hpp:18]: Yep, it changes `GPRF` state

File: src/exec/executor.cpp (hunk 4 above)
Comment (mmamayka) [reply src/exec/executor.cpp:42]: All known insns are implemented now

File: test/exec/simple-programs.cpp [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+    while (pc <= 20) {
+        exec.exec(instrs[pc / 4]);
+        pc = exec.getState().read(GPRF::PC);
+    }
+
+    EXPECT_EQ(exec.getState().read(GPRF::X1), 10);
+}
+
+/*
```
Comment (levBagryansky) [inline test/exec/simple-programs.cpp:50]: @mmamayka What is going here?

File: include/besm-666/exec/executor.hpp (hunk 5 above)
Comment (mmamayka) [reply include/besm-666/exec/executor.hpp]: Fixed

File: test/util/bit-magic-tests.cpp [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+TEST(SignExtend, PositiveValue) {
+    unsigned int v = 3; // 011
+
+    unsigned int r = besm::util::SignExtend<unsigned int, 3>(v);
+
+    EXPECT_EQ(r, 3u);
+}
+
+TEST(SignExtend, NegativeValue) {
```
Comment (levBagryansky) [inline test/util/bit-magic-tests.cpp:32]: @mmamayka please add the test where `uint v = 0b1111` and SignBits = 3

**Review by levBagryansky — CHANGES_REQUESTED** [review]
@mmamayka executor-tests is to review then)

File: test/util/bit-magic-tests.cpp (hunk 8 above)
Comment (mmamayka) [reply test/util/bit-magic-tests.cpp:32]: It is the same test as `NegativeValue`. Please, describe what corner case you want to be tested?

File: test/exec/simple-programs.cpp (hunk 7 above)
Comment (mmamayka) [reply test/exec/simple-programs.cpp:50]: The `EXPECT_EQ` checks that the result, placed in `x1` register, is equal to sum of range `0..5`.

File: test/util/bit-magic-tests.cpp (hunk 8 above)
Comment (levBagryansky) [reply test/util/bit-magic-tests.cpp:32]: In this test `7=0b111` with 3 '1' and SingnBits is `3` too. I'd like to be sure that the result with `0b1111` will be the same because 3 repeats in this test several times.

File: test/exec/simple-programs.cpp (hunk 7 above)
Comment (levBagryansky) [reply test/exec/simple-programs.cpp:50]: @mmamayka I don't get it. Could you leave a describing here? Maybe alternative "c" code in comment.

File: test/exec/RV64I-executor-tests.cpp [hunk 9]
```diff
@@ -0,0 +1,512 @@
...
+    }
+    void SetupInstrB(InstructionOp op, Register rs1, Register rs2,
+                     RV64UDWord imm) {
+        instr.operation = op;
+        instr.rs1 = rs1;
+        instr.rs2 = rs2;
+        instr.immidiate = imm;
+    }
+
+    void LoadImm12(Register rd, RV64DWord val) {
```
Comment (levBagryansky) [inline test/exec/RV64I-executor-tests.cpp:42]: @mmamayka Why 12? What does this function?

File: test/exec/RV64I-executor-tests.cpp [hunk 10]
```diff
@@ -0,0 +1,512 @@
...
+        instr.rs2 = rs2;
+    }
+    void SetupInstrI(InstructionOp op, Register rd, Register rs1,
+                     RV64UDWord imm) {
+        instr.operation = op;
+        instr.rd = rd;
+        instr.rs1 = rs1;
+        instr.immidiate = imm;
+    }
+    void SetupInstrIS(InstructionOp op, Register rd, Register rs1,
```
Comment (levBagryansky) [inline test/exec/RV64I-executor-tests.cpp:25]: @mmamayka `SetupInstrIS` is inappropriate name because `S` means either `format` and `sign`.

File: test/exec/RV64I-executor-tests.cpp (hunk 9 above)
Comment (mmamayka) [reply test/exec/RV64I-executor-tests.cpp:42]: It loads 12 bit wide immidiate value from `val` to register `rd`.

File: test/exec/RV64I-executor-tests.cpp (hunk 10 above)
Comment (mmamayka) [reply test/exec/RV64I-executor-tests.cpp:25]: The first capital letter `I` means I-format, the second capital `S` means signed

File: test/exec/simple-programs.cpp (hunk 7 above)
Comment (mmamayka) [reply test/exec/simple-programs.cpp:50]: There is a "c" alternative code in comment above the test

File: test/exec/RV64I-executor-tests.cpp (hunk 10 above)
Comment (levBagryansky) [reply test/exec/RV64I-executor-tests.cpp:25]: @mmamayka I get it and it is not clear as for me. Maybe `SetupInstrISigned`?

File: test/exec/simple-programs.cpp [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+        {GPRF::X3, GPRF::X0, GPRF::X0, 0, InstructionOp::ADD},
+        // PC = 12 - set up bitness to 32
+        {GPRF::X4, GPRF::X0, GPRF::X0, 32, InstructionOp::ADDI},
+        // PC = 16 - set up iterator to zero
+        {GPRF::X5, GPRF::X0, GPRF::X0, 0, InstructionOp::ADD},
+        // PC = 20 - get zero bit of second factor
+        {GPRF::X6, GPRF::X2, GPRF::X0, 1, InstructionOp::ANDI},
+        // PC = 24 - skip adding if bit is not set
+        {GPRF::X0, GPRF::X6, GPRF::X0, 4, InstructionOp::BEQ},
+        // PC = 28 - add first factor to result is bit is set
```
Comment (levBagryansky) [inline test/exec/simple-programs.cpp:74]: @mmamayka typo? `if bit is set`*?

File: test/util/bit-magic-tests.cpp (hunk 8 above)
Comment (levBagryansky) [reply test/util/bit-magic-tests.cpp:32]: @mmamayka up to you

File: test/exec/simple-programs.cpp (hunk 11 above)
Comment (mmamayka) [reply test/exec/simple-programs.cpp:74]: I do comparison with `x0` in `BEQ`, so it's not a typo.

File: test/exec/RV64I-executor-tests.cpp (hunk 10 above)
Comment (mmamayka) [reply test/exec/RV64I-executor-tests.cpp]: Fixed

File: test/util/bit-magic-tests.cpp (hunk 8 above)
Comment (mmamayka) [reply test/util/bit-magic-tests.cpp:32]: Done

**Review by levBagryansky — APPROVED** [review]
@mmamayka Thanks

**Review by c71n93 — APPROVED** [review]
